File: lib/parser.ts
/**
 * Styling options for text formatting.
 */
export enum StyleKind {
  /** Bold text `[b]` - renders as `<strong>` */
  BOLD = 'b',
  /** Italic text `[i]` - renders as `<em>` */
  ITALIC = 'i',
  /** Underlined text `[u]` - renders as `<ins>` */
  UNDERLINE = 'u',
  /** Strikethrough text `[s]` - renders as `<del>` */
  STRIKETHROUGH = 's',
  /** Link `[url]` - renders as `<a href="...">` */
  LINK = 'url',
}

/**
 * Emoticon types supported by sillycode.
 */
export enum EmoteKind {
  /** Smiley face `[:)]` - renders as: ![](https://sillypost.net/static/emoticons/smile.png) */
  SMILE = 'smile',
  /** Sad face `[:(]` - renders as: ![](https://sillypost.net/static/emoticons/sad.png) */
  SAD = 'sad',
  /** Big smile `[:D]` - renders as: ![](https://sillypost.net/static/emoticons/colond.png) */
  COLON_D = 'colond',
  /** Colon-three face `[:3]` - renders as: ![](https://sillypost.net/static/emoticons/colonthree.png) */
  COLON_THREE = 'colonthree',
  /** Fearful face `[D:]` - renders as: ![](https://sillypost.net/static/emoticons/fearful.png) */
  FEARFUL = 'fearful',
  /** Sunglasses `[B)]` - renders as: ![](https://sillypost.net/static/emoticons/sunglasses.png) */
  SUNGLASSES = 'sunglasses',
  /** Crying face `[;(]` - renders as: ![](https://sillypost.net/static/emoticons/crying.png) */
  CRYING = 'crying',
  /** Winking face `[;)]` - renders as: ![](https://sillypost.net/static/emoticons/winking.png) */
  WINKING = 'winking',
}

/** RGB color value with 8-bit components, represented as a hex string like `"#ad77f1"`. */
export type Color = string;

/** Regular text content part. */
export type TextPart = {
  type: 'text';
  text: string;
};

/** Escape backslash character part. */
export type EscapePart = {
  type: 'escape';
};

/** Line break character part. */
export type NewlinePart = {
  type: 'newline';
};

/** Style formatting toggle, enable or disable, each effect is independent. */
export type StylePart = {
  type: 'style';
  style: StyleKind;
  enable: boolean;
};

/** Color formatting toggle, enable or disable, acts as a stack. */
export type ColorPart = { type: 'color' } & (
  { enable: true, color: Color } |
  { enable: false }
);

/** Emoticon image part. */
export type EmotePart = {
  type: 'emote';
  emote: EmoteKind;
};

/**
 * A single element of parsed sillycode markup.
 */
export type Part =
  | TextPart
  | EscapePart
  | NewlinePart
  | StylePart
  | ColorPart
  | EmotePart;


/** lookup table for style tags like [b] or [/i] */
var styleParts = Object.create(null) as Record<string, StylePart>;

function addStyle(text: string, kind: StyleKind) {
  styleParts[      text] = { type: 'style', style: kind, enable: true  };
  styleParts['/' + text] = { type: 'style', style: kind, enable: false };
}

addStyle('b', StyleKind.BOLD);
addStyle('i', StyleKind.ITALIC);
addStyle('u', StyleKind.UNDERLINE);
addStyle('s', StyleKind.STRIKETHROUGH);
addStyle('url', StyleKind.LINK);

/** lookup table for emote tags like [:)] or [:D] */
var emoteParts = Object.create(null) as Record<string, EmotePart>;

function addEmote(text: string, kind: EmoteKind) {
  emoteParts[text] = { type: 'emote', emote: kind };
}

addEmote(':)', EmoteKind.SMILE);
addEmote(':(', EmoteKind.SAD);
addEmote(':D', EmoteKind.COLON_D);
addEmote(':3', EmoteKind.COLON_THREE);
addEmote('D:', EmoteKind.FEARFUL);
addEmote('B)', EmoteKind.SUNGLASSES);
addEmote(';(', EmoteKind.CRYING);
addEmote(';)', EmoteKind.WINKING);

/**
 * Parses sillycode markup into a list of parts.
 *
 * @param input The sillycode markup string to parse
 * @returns An array of parsed parts
 */
export function parse(input: string): Part[] {
  var parts: Part[] = [];

  // emits a new part
  function emit(part: Part): void {
    parts.push(part);
  }

  var buffer = '';

  // flushes the buffer as a text part if it's not empty
  function flush(): void {
    if (buffer) {
      emit({ type: 'text', text: buffer });
      buffer = '';
    }
  }

  // parses a tag body into a part
  function parseTag(body: string): Part | null {
    if (!body) {
      return null;
    }

    // check for style:
    var style = styleParts[body];
    if (style) {
      return style;
    }

    // check for emote:
    var emote = emoteParts[body];
    if (emote) {
      return emote;
    }

    // check for color push:
    var color = /^color=(#[0-9a-fA-F]{6})$/.exec(body);
    if (color) {
      return { type: 'color', enable: true, color: color[1]! };
    }

    // check for color pop:
    if (body === '/color') {
      return { type: 'color', enable: false };
    }

    return null;
  }

  // attempts to parse a tag at the current position
  function attemptTag(): boolean {
    // find the last opening bracket
    var index = buffer.lastIndexOf('[');
    if (index < 0) {
      return false;
    }

    // detect escape
    var lastPart = parts[parts.length - 1];
    if (lastPart && lastPart.type === 'escape' && index === 0) {
      return false;
    }

    // extract the tag body
    var body = buffer.slice(index + 1);

    // parse the tag
    var part = parseTag(body);

    // if we parsed a tag
    if (part) {
      // remove the tag from the buffer
      buffer = buffer.slice(0, index);
      // emit both the remaining buffer and the parsed part
      flush();
      emit(part);
      // success!
      return true;
    }

    // we did not parse a tag :<
    return false;
  }

  // is the next character escaped?
  var escape = false;

  // main parsing loop
  for (var i = 0; i < input.length; i++) {
    var char = input[i]!;

    // if we are not escaping
    if (!escape) {
      // check for escape
      if (char === '\\') {
        escape = true;
        flush();
        emit({ type: 'escape' });
        continue;
      }
      // check for tag close
      if (char === ']') {
        if (attemptTag()) {
          continue;
        }
      }
    }

    // make sure to reset the escape flag
    escape = false;

    // check for newline
    if (char === '\n') {
      flush();
      emit({ type: 'newline' });
      continue;
    }

    // collect normal characters in the buffer
    buffer += char;
  }

  // flush any remaining text
  flush();

  // we are done :3
  return parts;
}


/** counts the number of unicode scalars in a string */
function countUnicodeScalars(string: string): number {
  var count = 0;

  for (var i = 0; i < string.length; i++) {
    var code = string.charCodeAt(i);
    if (code >= 0xD800 && code <= 0xDBFF) {
      i++;
    }
    count++;
  }

  return count;
}


/**
 * Calculates the display length of parsed parts.
 *
 * @param parts The array of parsed parts to measure
 * @returns The display length as a number of visible characters
 */
export function length(parts: Part[]): number {
  var length = 0;

  parts.forEach(function (part) {
    if (part.type === 'text') {
      length += countUnicodeScalars(part.text);
    } else if (part.type === 'newline' || part.type === 'emote') {
      length += 1;
    }
  });

  return length;
}

File: src/parser.rs
use std::fmt;
use strum::IntoEnumIterator;
use strum_macros::EnumIter;

/// Styling options for text formatting.
#[derive(EnumIter, Debug, Clone, Copy, PartialEq)]
pub enum StyleKind {
  /// Bold text `[b]` - renders as `<strong>`
  Bold,
  /// Italic text `[i]` - renders as `<em>`
  Italic,
  /// Underlined text `[u]` - renders as `<ins>`
  Underline,
  /// Strikethrough text `[s]` - renders as `<del>`
  Strikethrough,
  /// Link `[url]` - renders as `<a href="...">`
  Link,
}

impl StyleKind {

  /// Returns the sillycode tag name for this style.
  pub const fn to_tag(&self) -> &str {
    match self {
      StyleKind::Bold => "b",
      StyleKind::Italic => "i",
      StyleKind::Underline => "u",
      StyleKind::Strikethrough => "s",
      StyleKind::Link => "url",
    }
  }

}

/// Emoticon types supported by sillycode.
#[derive(EnumIter, Debug, Clone, Copy, PartialEq)]
pub enum EmoteKind {
  /// Smiley face `[:)]` - renders as:
  /// ![](https://sillypost.net/static/emoticons/smile.png)
  Smile,
  /// Sad face `[:(`] - renders as:
  /// ![](https://sillypost.net/static/emoticons/sad.png)
  Sad,
  /// Big smile `[:D]` - renders as:
  /// ![](https://sillypost.net/static/emoticons/colond.png)
  ColonD,
  /// Colon-three face `[:3]` - renders as:
  /// ![](https://sillypost.net/static/emoticons/colonthree.png)
  ColonThree,
  /// Fearful face `[D:]` - renders as:
  /// ![](https://sillypost.net/static/emoticons/fearful.png)
  Fearful,
  /// Sunglasses `[B)]` - renders as:
  /// ![](https://sillypost.net/static/emoticons/sunglasses.png)
  Sunglasses,
  /// Crying face `[;(`] - renders as:
  /// ![](https://sillypost.net/static/emoticons/crying.png)
  Crying,
  /// Winking face `[;)]` - renders as:
  /// ![](https://sillypost.net/static/emoticons/winking.png)
  Winking,
}

impl EmoteKind {

  /// Returns the sillycode tag for this emoticon.
  pub const fn to_tag(&self) -> &str {
    match self {
      EmoteKind::Smile => ":)",
      EmoteKind::Sad => ":(",
      EmoteKind::ColonD => ":D",
      EmoteKind::ColonThree => ":3",
      EmoteKind::Fearful => "D:",
      EmoteKind::Sunglasses => "B)",
      EmoteKind::Crying => ";(",
      EmoteKind::Winking => ";)",
    }
  }

  /// Returns the file name for this emoticon without extension.
  pub const fn to_name(&self) -> &str {
    match self {
      EmoteKind::Smile => "smile",
      EmoteKind::Sad => "sad",
      EmoteKind::ColonD => "colond",
      EmoteKind::ColonThree => "colonthree",
      EmoteKind::Fearful => "fearful",
      EmoteKind::Sunglasses => "sunglasses",
      EmoteKind::Crying => "crying",
      EmoteKind::Winking => "winking",
    }
  }

}

/// RGB color value with 8-bit components.
#[derive(Default, Debug, Clone, Copy, PartialEq)]
pub struct Color {
  /// Red component (0-255).
  pub r: u8,
  /// Green component (0-255).
  pub g: u8,
  /// Blue component (0-255).
  pub b: u8,
}

impl Color {

  /// Creates a new color from RGB components.
  pub const fn new(r: u8, g: u8, b: u8) -> Self {
    Self { r, g, b }
  }

}

impl fmt::Display for Color {

  /// Formats the color as a hexadecimal string like `"#ad77f1"`.
  fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
    write!(f, "#{:02x}{:02x}{:02x}", self.r, self.g, self.b)
  }

}

/// A single element of parsed sillycode markup.
#[derive(Debug, Clone, PartialEq)]
pub enum Part {
  /// Plain text content.
  Text(String),
  /// Escape backslash character.
  Escape,
  /// Line break character.
  Newline,
  /// Style formatting toggle, enable or disable, each effect is independent.
  Style(StyleKind, bool),
  /// Color formatting toggle, enable or disable, acts as a stack.
  Color(Color, bool),
  /// Emoticon image.
  Emote(EmoteKind),
}

impl Part {

  /// parses a style tag body like "b" or "/url"
  fn parse_style_tag(mut body: &str) -> Option<Self> {
    let mut enable = true;

    // check if the tag is closing
    if body.starts_with('/') {
      enable = false;
      body = &body[1..];
    }

    for style in StyleKind::iter() {
      if body == style.to_tag() {
        return Some(Self::Style(style, enable));
      }
    }

    None
  }

  /// parses an emote tag body like ":)"
  fn parse_emote_tag(body: &str) -> Option<Self> {
    for emote in EmoteKind::iter() {
      if body == emote.to_tag() {
        return Some(Self::Emote(emote));
      }
    }

    None
  }

  /// parses a color tag body like "color=#ad77f1"
  fn parse_color_tag(body: &str) -> Option<Self> {
    if body.len() == 13 && body.starts_with("color=#") {
      let r = u8::from_str_radix(&body[ 7.. 9], 16).ok()?;
      let g = u8::from_str_radix(&body[ 9..11], 16).ok()?;
      let b = u8::from_str_radix(&body[11..13], 16).ok()?;
      Some(Self::Color(Color::new(r, g, b), true))
    } else if body == "/color" {
      Some(Self::Color(Color::default(), false))
    } else {
      None
    }
  }

  /// parses any tag body
  fn parse_tag(body: &str) -> Option<Self> {
    if body.is_empty() || body.len() > 32 {
      return None;
    }

    Self::parse_style_tag(body)
      .or_else(|| Self::parse_emote_tag(body))
      .or_else(|| Self::parse_color_tag(body))
  }

}

impl fmt::Display for Part {

  /// Formats the part back to sillycode markup.
  fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
    match self {
      Part::Text(text) => write!(f, "{text}"),
      Part::Escape => write!(f, "\\"),
      Part::Newline => write!(f, "\n"),
      Part::Style(style, enable) => {
        if *enable {
          write!(f, "[{}]", style.to_tag())
        } else {
          write!(f, "[/{}]", style.to_tag())
        }
      }
      Part::Color(color, enable) => {
        if *enable {
          write!(f, "[color={}]", color)
        } else {
          write!(f, "[/color]")
        }
      }
      Part::Emote(emote) => write!(f, "[{}]", emote.to_tag()),
    }
  }

}

/// parser for sillycode markup
#[derive(Default, Debug)]
struct Parser {
  /// output parts
  parts: Vec<Part>,
  /// buffer for text parts
  buffer: String,
  /// whether the previous character was an escape
  escape: bool,
}

impl Parser {

  /// creates a new parser
  fn new() -> Self {
    Self::default()
  }

  /// emits a new part
  fn emit(&mut self, part: Part) {
    self.parts.push(part);
  }

  /// flushes the buffer as a text part if it's not empty
  fn flush(&mut self) {
    if !self.buffer.is_empty() {
      self.emit(Part::Text(self.buffer.clone()));
      self.buffer.clear();
    }
  }

  /// attempts to parse a tag at the current position
  fn tag(&mut self) -> bool {
    // find the last opening bracket
    let index = match self.buffer.rfind('[') {
        Some(index) => index,
        None => return false,
    };

    // detect escape
    if index == 0 && matches!(self.parts.last(), Some(Part::Escape)) {
      return false;
    }

    // extract the tag body
    let body = &self.buffer[index+1..];

    // parse the tag
    let part = Part::parse_tag(body);

    // if we parsed a tag
    if let Some(part) = part {
      // remove the tag from the buffer
      self.buffer.drain(index..);
      // emit both the remaining buffer and the parsed part
      self.flush();
      self.emit(part);
      // success!
      true
    } else {
      // we did not parse a tag :<
      false
    }
  }

  /// parses sillycode markup
  fn parse(mut self, input: &str) -> Vec<Part> {
    // main parsing loop
    for char in input.chars() {
      // if we are not escaping
      if !self.escape {
        // check for escape
        if char == '\\' {
          self.escape = true;
          self.flush();
          self.emit(Part::Escape);
          continue;
        }
        // check for tag close
        if char == ']' {
          if self.tag() {
            continue;
          }
        }
      }

      // make sure to reset the escape flag
      self.escape = false;

      // check for newline
      if char == '\n' {
        self.flush();
        self.emit(Part::Newline);
        continue;
      }

      // collect normal characters in the buffer
      self.buffer.push(char);
    }

    // flush any remaining text
    self.flush();

    // we are done :3
    self.parts
  }

}

/// Parses sillycode markup into a list of parts.
pub fn parse(input: &str) -> Vec<Part> {
  Parser::new().parse(input)
}

/// Calculates the display length of parsed parts.
pub fn length(parts: &[Part]) -> usize {
  parts.iter().fold(0, |acc, part| {
    match part {
      Part::Text(text) => acc + text.chars().count(),
      Part::Newline | Part::Emote(_) => acc + 1,
      _ => acc,
    }
  })
}
